File: backend/core/src/modules/translations/translations.service.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { ProjectAccessService } from '../../common/access/project-access.service';
import { Document, Translation } from '../../entities';
import { AiService } from '../ai/ai.service';

@Injectable()
export class TranslationsService {
  constructor(
    @InjectRepository(Translation)
    private readonly translationsRepo: Repository<Translation>,
    @InjectRepository(Document)
    private readonly documentsRepo: Repository<Document>,
    private readonly projectAccessService: ProjectAccessService,
    private readonly aiService: AiService,
  ) {}

  private async assertDocumentAccess(userId: string, documentId: string): Promise<Document> {
    const doc = await this.documentsRepo.findOne({ where: { id: documentId } });
    if (!doc) {
      throw new BadRequestException('문서를 찾을 수 없습니다.');
    }
    await this.projectAccessService.assertProjectAccess(userId, doc.projectId);
    return doc;
  }

  async findAllForUser(userId: string, documentId: string) {
    await this.assertDocumentAccess(userId, documentId);
    return this.translationsRepo.find({
      where: { documentId } as any,
      order: { createdAt: 'DESC' } as any,
    });
  }

  async findOneForUser(userId: string, id: string) {
    const item = await this.translationsRepo.findOne({ where: { id } });
    if (!item) return null;
    await this.assertDocumentAccess(userId, item.documentId);
    return item;
  }

  async createForUser(userId: string, dto: Partial<Translation>) {
    if (!dto.documentId) {
      throw new BadRequestException('documentId가 필요합니다.');
    }
    await this.assertDocumentAccess(userId, dto.documentId);
    const entity = this.translationsRepo.create(dto as any);
    return this.translationsRepo.save(entity as any);
  }

  async generateForUser(
    user: { userId: string; role?: string },
    dto: { documentId?: string; targetLanguage?: string; provider?: string; model?: string },
  ) {
    if (!dto.documentId) {
      throw new BadRequestException('documentId가 필요합니다.');
    }
    if (!dto.targetLanguage) {
      throw new BadRequestException('targetLanguage가 필요합니다.');
    }

    const doc = await this.assertDocumentAccess(user.userId, dto.documentId);
    const text = [doc.title, doc.content].filter(Boolean).join('\n\n');
    if (!text) {
      throw new BadRequestException('문서 내용이 없습니다.');
    }

    const ai = await this.aiService.proxy(user, 'translate', '/ai/translate', {
      text,
      target_language: dto.targetLanguage,
      provider: dto.provider,
      model: dto.model,
      projectId: doc.projectId,
    });

    const entity = this.translationsRepo.create({
      documentId: doc.id,
      targetLanguage: dto.targetLanguage,
      provider: dto.provider,
      content: typeof ai?.content === 'string' ? ai.content : undefined,
    } as any);
    return this.translationsRepo.save(entity as any);
  }

  async removeForUser(userId: string, id: string) {
    const entity = await this.translationsRepo.findOne({ where: { id } });
    if (!entity) return;
    await this.assertDocumentAccess(userId, entity.documentId);
    await this.translationsRepo.delete({ id } as any);
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/TranslationsService.java
package store.pjcloud.cowrite.core.service;

import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;
import store.pjcloud.cowrite.core.entity.Document;
import store.pjcloud.cowrite.core.entity.Translation;
import store.pjcloud.cowrite.core.repository.DocumentRepository;
import store.pjcloud.cowrite.core.repository.TranslationRepository;

@Service
public class TranslationsService {
    private final TranslationRepository translationsRepo;
    private final DocumentRepository documentsRepo;
    private final ProjectAccessService projectAccessService;
    private final AiService aiService;

    public TranslationsService(TranslationRepository translationsRepo,
                               DocumentRepository documentsRepo,
                               ProjectAccessService projectAccessService,
                               AiService aiService) {
        this.translationsRepo = translationsRepo;
        this.documentsRepo = documentsRepo;
        this.projectAccessService = projectAccessService;
        this.aiService = aiService;
    }

    private Document assertDocumentAccess(UUID userId, UUID documentId) {
        Document doc = documentsRepo.findById(documentId).orElse(null);
        if (doc == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "문서를 찾을 수 없습니다.");
        projectAccessService.assertProjectAccess(userId, doc.getProjectId());
        return doc;
    }

    public List<Translation> findAllForUser(UUID userId, UUID documentId) {
        assertDocumentAccess(userId, documentId);
        return translationsRepo.findByDocumentIdOrderByCreatedAtDesc(documentId);
    }

    public Translation findOneForUser(UUID userId, UUID id) {
        Translation entity = translationsRepo.findById(id).orElse(null);
        if (entity == null) return null;
        assertDocumentAccess(userId, entity.getDocumentId());
        return entity;
    }

    @Transactional
    public Translation createForUser(UUID userId, Translation dto) {
        if (dto.getDocumentId() == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "documentId가 필요합니다.");
        assertDocumentAccess(userId, dto.getDocumentId());
        return translationsRepo.save(dto);
    }

    @Transactional
    public Translation generateForUser(AiService.UserContext user, Map<String, Object> body) {
        Object documentIdObj = body.get("documentId");
        Object targetLangObj = body.get("targetLanguage");
        if (!(documentIdObj instanceof String) || ((String) documentIdObj).isBlank()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "documentId가 필요합니다.");
        }
        if (!(targetLangObj instanceof String) || ((String) targetLangObj).isBlank()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "targetLanguage가 필요합니다.");
        }

        UUID documentId = UUID.fromString((String) documentIdObj);
        Document doc = assertDocumentAccess(user.userId(), documentId);
        String text = String.join("\n\n", java.util.List.of(doc.getTitle(), doc.getContent()).stream().filter(v -> v != null && !v.isBlank()).toList());
        if (text.isBlank()) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "문서 내용이 없습니다.");

        Map<String, Object> payload = Map.of(
            "text", text,
            "target_language", targetLangObj,
            "provider", body.get("provider"),
            "model", body.get("model"),
            "projectId", doc.getProjectId().toString()
        );

        Object ai = aiService.proxy(user, "translate", "/ai/translate", new java.util.HashMap<>(payload));
        String content = null;
        if (ai instanceof Map<?, ?> map && map.get("content") instanceof String s) content = s;

        Translation entity = new Translation();
        entity.setDocumentId(doc.getId());
        entity.setTargetLanguage((String) targetLangObj);
        entity.setProvider(body.get("provider") instanceof String s ? s : null);
        entity.setContent(content);
        return translationsRepo.save(entity);
    }

    @Transactional
    public void removeForUser(UUID userId, UUID id) {
        Translation entity = translationsRepo.findById(id).orElse(null);
        if (entity == null) return;
        assertDocumentAccess(userId, entity.getDocumentId());
        translationsRepo.deleteById(id);
    }
}
